diff --git a/scripts/package/mkdebian b/scripts/package/mkdebian
index fc3b7fa70..8ff6607f8 100755
--- a/scripts/package/mkdebian
+++ b/scripts/package/mkdebian
@@ -201,7 +201,7 @@ Build-Depends: debhelper-compat (= 12)
 Build-Depends-Arch: bc, bison, cpio, flex,
  gcc-${host_gnu} <!pkg.${sourcename}.nokernelheaders>,
  kmod, libelf-dev:native,
- libssl-dev:native, libssl-dev <!pkg.${sourcename}.nokernelheaders>,
+ libssl-dev:native,
  rsync
 Homepage: https://www.kernel.org/
 
