--- linux-6.12.10/kernel/irq/handle.c	2025-07-02 11:58:57.337777200 +0900
+++ linux-xlnx-2025.1/kernel/irq/handle.c	2025-07-02 12:01:41.071761200 +0900
@@ -215,7 +215,11 @@
 }
 
 #ifdef CONFIG_GENERIC_IRQ_MULTI_HANDLER
+#ifndef CONFIG_IRQCHIP_XILINX_INTC_MODULE_SUPPORT_EXPERIMENTAL
 int __init set_handle_irq(void (*handle_irq)(struct pt_regs *))
+#else
+int set_handle_irq(void (*handle_irq)(struct pt_regs *))
+#endif
 {
 	if (handle_arch_irq)
 		return -EBUSY;
