--- linux-6.12.10/net/ipv4/af_inet.c	2025-07-02 11:58:58.078300200 +0900
+++ linux-xlnx-2025.1/net/ipv4/af_inet.c	2025-07-02 12:01:41.811425600 +0900
@@ -122,6 +122,7 @@
 #include <net/rps.h>
 
 #include <trace/events/sock.h>
+#include <xen/pvcalls.h>
 
 /* The inetsw table contains everything that inet_create needs to
  * build a new socket.
@@ -1965,6 +1966,11 @@
 	for (r = &inetsw[0]; r < &inetsw[SOCK_MAX]; ++r)
 		INIT_LIST_HEAD(r);
 
+	if (pvcalls) {
+		pr_info("Enabling pvcalls for AF_INET SOCK_STREAM\n");
+		inetsw_array[0].ops = &pvcalls_stream_ops;
+	}
+
 	for (q = inetsw_array; q < &inetsw_array[INETSW_ARRAY_LEN]; ++q)
 		inet_register_protosw(q);
 
