--- linux-6.12.10/crypto/ecdsa.c	2025-07-02 11:58:37.849781600 +0900
+++ linux-xlnx-2025.1/crypto/ecdsa.c	2025-07-02 12:01:21.460374300 +0900
@@ -8,6 +8,7 @@
 #include <crypto/internal/ecc.h>
 #include <crypto/akcipher.h>
 #include <crypto/ecdh.h>
+#include <crypto/ecdsa.h>
 #include <linux/asn1_decoder.h>
 #include <linux/scatterlist.h>
 
@@ -23,17 +24,11 @@
 	struct ecc_point pub_key;
 };
 
-struct ecdsa_signature_ctx {
-	const struct ecc_curve *curve;
-	u64 r[ECC_MAX_DIGITS];
-	u64 s[ECC_MAX_DIGITS];
-};
-
 /*
  * Get the r and s components of a signature from the X509 certificate.
  */
-static int ecdsa_get_signature_rs(u64 *dest, size_t hdrlen, unsigned char tag,
-				  const void *value, size_t vlen, unsigned int ndigits)
+int ecdsa_get_signature_rs(u64 *dest, size_t hdrlen, unsigned char tag,
+			   const void *value, size_t vlen, unsigned int ndigits)
 {
 	size_t bufsize = ndigits * sizeof(u64);
 	const char *d = value;
@@ -59,6 +54,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL(ecdsa_get_signature_rs);
 
 int ecdsa_get_signature_r(void *context, size_t hdrlen, unsigned char tag,
 			  const void *value, size_t vlen)
