--- linux-6.12.10/drivers/clocksource/timer-cadence-ttc.c	2025-07-02 11:58:39.097752600 +0900
+++ linux-xlnx-2025.1/drivers/clocksource/timer-cadence-ttc.c	2025-07-02 12:01:22.710243300 +0900
@@ -489,6 +489,13 @@
 	u32 timer_width = 16;
 	struct device_node *timer = pdev->dev.of_node;
 
+	/*
+	 * If pwm-cells property is present in TTC node,
+	 * it would be treated as PWM device.
+	 */
+	if (of_property_read_bool(timer, "#pwm-cells"))
+		return -ENODEV;
+
 	if (initialized)
 		return 0;
 
