--- linux-6.12.10/Documentation/devicetree/bindings/mailbox/xlnx,zynqmp-ipi-mailbox.yaml	2025-07-02 11:58:27.288233800 +0900
+++ linux-xlnx-2025.1/Documentation/devicetree/bindings/mailbox/xlnx,zynqmp-ipi-mailbox.yaml	2025-07-02 12:01:12.052860000 +0900
@@ -61,9 +61,11 @@
     const: 2
 
   reg:
+    minItems: 1
     maxItems: 2
 
   reg-names:
+    minItems: 1
     maxItems: 2
 
   xlnx,ipi-id:
@@ -167,11 +169,13 @@
     else:
       properties:
         reg:
+          minItems: 1
           items:
             - description: Host IPI agent control register region
             - description: Host IPI agent optional message buffers
 
         reg-names:
+          minItems: 1
           items:
             - const: ctrl
             - const: msg
--- linux-6.12.10/drivers/mailbox/zynqmp-ipi-mailbox.c	2025-07-02 11:58:45.301842700 +0900
+++ linux-xlnx-2025.1/drivers/mailbox/zynqmp-ipi-mailbox.c	2025-07-02 12:01:29.099261300 +0900
@@ -905,7 +905,7 @@
 {
 	struct device *dev = &pdev->dev;
 	struct device_node *nc, *np = pdev->dev.of_node;
-	struct zynqmp_ipi_pdata __percpu *pdata;
+	struct zynqmp_ipi_pdata *pdata;
 	struct of_phandle_args out_irq;
 	struct zynqmp_ipi_mbox *mbox;
 	int num_mboxes, ret = -EINVAL;
@@ -940,10 +940,10 @@
 	pdata->num_mboxes = num_mboxes;
 
 	mbox = pdata->ipi_mboxes;
-	mbox->setup_ipi_fn = ipi_fn;
-
 	for_each_available_child_of_node(np, nc) {
 		mbox->pdata = pdata;
+		mbox->setup_ipi_fn = ipi_fn;
+
 		ret = zynqmp_ipi_mbox_probe(mbox, nc);
 		if (ret) {
 			of_node_put(nc);
