--- linux-6.12.10/drivers/rtc/rtc-zynqmp.c	2025-07-02 11:58:51.120303100 +0900
+++ linux-xlnx-2025.1/drivers/rtc/rtc-zynqmp.c	2025-07-02 12:01:34.985842600 +0900
@@ -318,8 +318,8 @@
 		return ret;
 	}
 
-	/* Getting the rtc_clk info */
-	xrtcdev->rtc_clk = devm_clk_get_optional(&pdev->dev, "rtc_clk");
+	/* Getting the rtc info */
+	xrtcdev->rtc_clk = devm_clk_get_optional(&pdev->dev, "rtc");
 	if (IS_ERR(xrtcdev->rtc_clk)) {
 		if (PTR_ERR(xrtcdev->rtc_clk) != -EPROBE_DEFER)
 			dev_warn(&pdev->dev, "Device clock not found.\n");
