--- linux-6.12.10/Documentation/devicetree/bindings/soc/xilinx/xilinx.yaml	2025-07-02 11:58:28.109845400 +0900
+++ linux-xlnx-2025.1/Documentation/devicetree/bindings/soc/xilinx/xilinx.yaml	2025-07-02 12:01:13.011919100 +0900
@@ -10,7 +10,7 @@
   - Michal Simek <michal.simek@amd.com>
 
 description: |
-  Xilinx boards with Zynq-7000 SOC or Zynq UltraScale+ MPSoC
+  Xilinx boards with Zynq-7000 SOC or Zynq UltraScale+ MPSoC or Versal or Versal NET
 
 properties:
   $nodename:
@@ -63,7 +63,10 @@
 
       - description: Xilinx evaluation board zcu1275
         items:
-          - const: xlnx,zynqmp-zcu1275-revA
+          - enum:
+              - xlnx,zynqmp-zcu1275-revA
+              - xlnx,zynqmp-zcu1275-revB
+              - xlnx,zynqmp-zcu1275-rev1.0
           - const: xlnx,zynqmp-zcu1275
           - const: xlnx,zynqmp
 
@@ -81,6 +84,14 @@
           - const: xlnx,zynqmp-zcu100
           - const: xlnx,zynqmp
 
+      - description: Xilinx ZynqMP based System Controller
+        items:
+          - enum:
+              - xlnx,zynqmp-sc-revB
+              - xlnx,zynqmp-sc-revC
+          - const: xlnx,zynqmp-sc
+          - const: xlnx,zynqmp
+
       - description: Xilinx evaluation board zcu102
         items:
           - enum:
@@ -116,6 +127,29 @@
           - const: xlnx,zynqmp-zcu111
           - const: xlnx,zynqmp
 
+      - description: Xilinx evaluation board zcu208
+        items:
+          - enum:
+              - xlnx,zynqmp-zcu208-revA
+              - xlnx,zynqmp-zcu208-rev1.0
+          - const: xlnx,zynqmp-zcu208
+          - const: xlnx,zynqmp
+
+      - description: Xilinx evaluation board zcu216
+        items:
+          - enum:
+              - xlnx,zynqmp-zcu216-revA
+              - xlnx,zynqmp-zcu216-rev1.0
+          - const: xlnx,zynqmp-zcu216
+          - const: xlnx,zynqmp
+
+      - description: Xilinx evaluation board zcu1285
+        items:
+          - enum:
+              - xlnx,zynqmp-zcu1285-revA
+              - xlnx,zynqmp-zcu1285-rev1.0
+          - const: xlnx,zynqmp-zcu1285
+
       - description: Xilinx Kria SOMs
         minItems: 3
         items:
@@ -182,6 +216,12 @@
           - contains:
               const: xlnx,zynqmp
 
+      - description: Xilinx Versal NET b2197 (Tenzing2)
+        items:
+          - const: xlnx,versal-net-vn-p-b2197-00-revA
+          - const: xlnx,versal-net-vn-p-b2197-00
+          - const: xlnx,versal-net
+
       - description: AMD MicroBlaze V (QEMU)
         items:
           - const: qemu,mbv
--- /dev/null
+++ linux-xlnx-2025.1/Documentation/devicetree/bindings/soc/xilinx/xlnx,ai_engine.txt	2025-07-02 12:01:13.012916300 +0900
@@ -0,0 +1,28 @@
+Xilinx AI Engine
+----------------
+
+The Xilinx AI Engine is a tile processor with many cores (up to 400) that
+can run in parallel. The data routing between cores is configured through
+internal switches, and shim tiles interface with external interconnect, such
+as memory or PL.
+
+Required properties:
+
+- compatible: Must be "xlnx,ai_engine".
+- reg: Physical base address and length of the registers set for the device.
+- interrupt-parent: the phandle to the interrupt controller.
+- interrupts: the interrupt numbers.
+- interrupt-names: Should be "interrupt0", "interrupt1", "interrupt2" or
+  "interrupt3".
+
+Example:
+
+	ai_engine@20000000000 {
+		compatible = "xlnx,ai_engine";
+		reg = <0x200 0x0 0x1 0x0>;
+		interrupt-parent = <&gic>;
+		interrupts = <0x0 0x94 0x1>,
+			     <0x0 0x95 0x1>,
+			     <0x0 0x96 0x1>;
+		interrupt-names = "interrupt1", "interrupt2", "interrupt3";
+	};
--- linux-6.12.10/drivers/soc/xilinx/xlnx_event_manager.c	2025-07-02 11:58:51.877789700 +0900
+++ linux-xlnx-2025.1/drivers/soc/xilinx/xlnx_event_manager.c	2025-07-02 12:01:35.730145600 +0900
@@ -3,12 +3,15 @@
  * Xilinx Event Management Driver
  *
  *  Copyright (C) 2021 Xilinx, Inc.
- *  Copyright (C) 2024 Advanced Micro Devices, Inc.
+ *  Copyright (C) 2024-2025 Advanced Micro Devices, Inc.
  *
  *  Abhyuday Godhasara <abhyuday.godhasara@xilinx.com>
  */
 
 #include <linux/cpuhotplug.h>
+#include <linux/firmware/xlnx-versal-error-events.h>
+#include <linux/firmware/xlnx-versal-net-error-events.h>
+#include <linux/firmware/amd-versal2-error-events.h>
 #include <linux/firmware/xlnx-event-manager.h>
 #include <linux/firmware/xlnx-zynqmp.h>
 #include <linux/hashtable.h>
@@ -81,11 +84,12 @@
 
 	zynqmp_pm_get_family_info(&pm_family_code, &pm_sub_family_code);
 
-	if (pm_sub_family_code == VERSAL_SUB_FAMILY_CODE) {
+	if (pm_sub_family_code <= VERSAL_SUB_FAMILY_CODE_MAX) {
 		if (node_id == VERSAL_EVENT_ERROR_PMC_ERR1 ||
 		    node_id == VERSAL_EVENT_ERROR_PMC_ERR2 ||
 		    node_id == VERSAL_EVENT_ERROR_PSM_ERR1 ||
-		    node_id == VERSAL_EVENT_ERROR_PSM_ERR2)
+		    node_id == VERSAL_EVENT_ERROR_PSM_ERR2 ||
+		    node_id == VERSAL_EVENT_ERROR_SW_ERR)
 			return true;
 	} else {
 		if (node_id == VERSAL_NET_EVENT_ERROR_PMC_ERR1 ||
@@ -94,7 +98,16 @@
 		    node_id == VERSAL_NET_EVENT_ERROR_PSM_ERR1 ||
 		    node_id == VERSAL_NET_EVENT_ERROR_PSM_ERR2 ||
 		    node_id == VERSAL_NET_EVENT_ERROR_PSM_ERR3 ||
-		    node_id == VERSAL_NET_EVENT_ERROR_PSM_ERR4)
+		    node_id == VERSAL_NET_EVENT_ERROR_PSM_ERR4 ||
+		    node_id == VERSAL_NET_EVENT_ERROR_SW_ERR ||
+		    node_id == VERSAL2_EVENT_ERROR_PMC_ERR1 ||
+		    node_id == VERSAL2_EVENT_ERROR_PMC_ERR2 ||
+		    node_id == VERSAL2_EVENT_ERROR_PMC_ERR3 ||
+		    node_id == VERSAL2_EVENT_ERROR_LPDSLCR_ERR1 ||
+		    node_id == VERSAL2_EVENT_ERROR_LPDSLCR_ERR2 ||
+		    node_id == VERSAL2_EVENT_ERROR_LPDSLCR_ERR3 ||
+		    node_id == VERSAL2_EVENT_ERROR_LPDSLCR_ERR4 ||
+		    node_id == VERSAL2_EVENT_ERROR_SW_ERR)
 			return true;
 	}
 
--- linux-6.12.10/drivers/soc/xilinx/zynqmp_power.c	2025-07-02 11:58:51.878853600 +0900
+++ linux-xlnx-2025.1/drivers/soc/xilinx/zynqmp_power.c	2025-07-02 12:01:35.731172900 +0900
@@ -319,7 +319,7 @@
 		if (ret < 0)
 			return ret;
 
-		if (pm_sub_family_code == VERSALNET_SUB_FAMILY_CODE)
+		if (pm_sub_family_code == VERSALNET_OR_VERSAL2_SUB_FAMILY_CODE)
 			node_id = PM_DEV_ACPU_0_0;
 		else
 			node_id = PM_DEV_ACPU_0;
