--- linux-6.12.10/drivers/virtio/virtio_ring.c	2025-07-02 11:58:53.494617400 +0900
+++ linux-xlnx-2025.1/drivers/virtio/virtio_ring.c	2025-07-02 12:01:37.369662000 +0900
@@ -291,7 +291,7 @@
 	 * the DMA API if we're a Xen guest, which at least allows
 	 * all of the sensible Xen configurations to work correctly.
 	 */
-	if (xen_domain())
+	if (xen_vring_use_dma())
 		return true;
 
 	return false;
